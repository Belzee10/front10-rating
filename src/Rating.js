import React from 'react';
import PropTypes from 'prop-types';
import '../assets/style.css';

import Item from './Item';

const renderItemType = (index, value, renderFullItem, renderEmptyItem) => {
  let icon = '';
  if (index <= value) {
    if (renderFullItem && typeof renderFullItem === 'function')
      return renderFullItem(index);
    icon += 'icon-star-full';
  } else {
    if (renderEmptyItem && typeof renderEmptyItem === 'function')
      return renderEmptyItem(index);
    icon += 'icon-star-empty';
  }
  return <Item key={index} icon={icon} />;
};

const renderItems = ({
  starsLength,
  value,
  renderItem,
  renderFullItem,
  renderEmptyItem
  // ...props
}) => {
  let items = [];
  for (let index = 0; index < starsLength; index += 1) {
    const itemToRender =
      renderItem && typeof renderItem === 'function'
        ? renderItem(index)
        : renderItemType(index + 1, value, renderFullItem, renderEmptyItem);

    items = [...items, itemToRender];
  }
  return items;
};

const Rating = props => {
  return renderItems(props);
};

Rating.propTypes = {
  /**
   * Total of items to show
   */
  starsLength: PropTypes.number,
  /**
   * Rating value
   */
  value: PropTypes.number,
  /**
   * Function to render the Items
   */
  renderItem: PropTypes.func,
  /**
   * Function to render the marked items
   */
  renderFullItem: PropTypes.func,
  /**
   * Function to render the unmarked Items
   */
  renderEmptyItem: PropTypes.func
};

Rating.defaultProps = {
  starsLength: 5,
  value: 0,
  renderItem: null,
  renderFullItem: null,
  renderEmptyItem: null
};

export default Rating;
