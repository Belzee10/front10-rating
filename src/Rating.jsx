import React from 'react';
import PropTypes from 'prop-types';
import '../assets/style.css';

import Item from './Item';

const renderItemType = (
  allowRate,
  index,
  value,
  renderFullItem,
  renderEmptyItem,
  renderHalfItem,
  onKeyPress,
  onClick
) => {
  let icon = '';
  const roundedValue = Math.ceil(value);
  if (index <= roundedValue) {
    if (index === roundedValue && value !== roundedValue) {
      if (renderHalfItem && typeof renderHalfItem === 'function') return renderHalfItem(index);
      icon += 'icon-star-half';
    } else {
      if (renderFullItem && typeof renderFullItem === 'function') return renderFullItem(index);
      icon = 'icon-star-full';
    }
  } else {
    if (renderEmptyItem && typeof renderEmptyItem === 'function') return renderEmptyItem(index);
    icon = 'icon-star-empty';
  }
  return (
    <Item
      key={index}
      icon={icon}
      allowRate={allowRate}
      onClick={onClick(index)}
      onKeyPress={onKeyPress(index)}
    />
  );
};

const Rating = ({
  allowRate,
  starsLength,
  value,
  renderItem,
  renderFullItem,
  renderEmptyItem,
  renderHalfItem,
  onKeyPress,
  onClick,
  className,
  style
}) => {
  let items = [];
  for (let index = 0; index < starsLength; index += 1) {
    const itemToRender =
      renderItem && typeof renderItem === 'function'
        ? renderItem(index)
        : renderItemType(
            allowRate,
            index + 1,
            value,
            renderFullItem,
            renderEmptyItem,
            renderHalfItem,
            onKeyPress,
            onClick
          );
    items = [...items, itemToRender];
  }
  return (
    <div className={`fr-rating-container ${className}`} style={style}>
      {items}
    </div>
  );
};

Rating.propTypes = {
  /**
   * Allow to rate
   */
  allowRate: PropTypes.bool,
  /**
   * Total of items to show
   */
  starsLength: PropTypes.number,
  /**
   * Rating value
   */
  value: PropTypes.number,
  /**
   * Function to render the Items
   */
  renderItem: PropTypes.func,
  /**
   * Function to render the marked items
   */
  renderFullItem: PropTypes.func,
  /**
   * Function to render the unmarked Items
   */
  renderEmptyItem: PropTypes.func,
  /**
   * Function to render the half items
   */
  renderHalfItem: PropTypes.func,
  /**
   * Function to handle rate on click
   */
  onKeyPress: PropTypes.func,
  /**
   * Function to handle rate on key press
   */
  onClick: PropTypes.func,
  /**
   * className applied to the component
   */
  className: PropTypes.string,
  /**
   * inline styles applied to the component
   */
  style: PropTypes.objectOf(PropTypes.any)
};

Rating.defaultProps = {
  allowRate: false,
  starsLength: 5,
  value: 0,
  renderItem: null,
  renderFullItem: null,
  renderEmptyItem: null,
  renderHalfItem: null,
  onKeyPress: null,
  onClick: null,
  className: '',
  style: null
};

export default Rating;
