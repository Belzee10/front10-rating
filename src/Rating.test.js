import React from 'react';
import { render, cleanup, waitForElement, fireEvent } from '@testing-library/react';
import 'jest-dom/extend-expect';

import Rating from './Rating';

afterEach(cleanup);

test('Testing Basic', async () => {
  const { container, asFragment } = render(<Rating starsLength={5} value={3} />);

  expect(container.firstChild).toMatchSnapshot();
  expect(asFragment()).toMatchSnapshot();
});
test('Testing functionalities', async () => {
  const onKeyPress = jest.fn();
  const onClick = jest.fn();
  const { getAllByRole, container, asFragment } = render(
    <Rating allowRate starsLength={5} value={3} onKeyPress={onKeyPress} onClick={onClick} />
  );
  const items = await waitForElement(() => getAllByRole('button'));
  expect(items.length).toBe(5);

  fireEvent.click(items[3]);
  fireEvent.keyPress(items[3], { keyCode: 13 });
  expect(onKeyPress).toHaveBeenCalled();
  expect(onClick).toHaveBeenCalled();
  // snapshots work great with regular DOM nodes!
  expect(container.firstChild).toMatchSnapshot();
  // you can also get a `DocumentFragment`, which is useful if you want to compare nodes across renders
  expect(asFragment()).toMatchSnapshot();
});
