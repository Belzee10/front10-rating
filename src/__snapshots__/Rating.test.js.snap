// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Basic 1`] = `
<div
  class="fr-rating-container "
>
  <i
    class="fr-rating-item icon-star-full"
  />
  <i
    class="fr-rating-item icon-star-full"
  />
  <i
    class="fr-rating-item icon-star-full"
  />
  <i
    class="fr-rating-item icon-star-empty"
  />
  <i
    class="fr-rating-item icon-star-empty"
  />
</div>
`;

exports[`Testing Basic 2`] = `
<DocumentFragment>
  <div
    class="fr-rating-container "
  >
    <i
      class="fr-rating-item icon-star-full"
    />
    <i
      class="fr-rating-item icon-star-full"
    />
    <i
      class="fr-rating-item icon-star-full"
    />
    <i
      class="fr-rating-item icon-star-empty"
    />
    <i
      class="fr-rating-item icon-star-empty"
    />
  </div>
</DocumentFragment>
`;

exports[`Testing functionalities 1`] = `
<div
  class="fr-rating-container "
>
  <i
    class="fr-rating-item icon-star-full"
    role="button"
    tabindex="0"
  />
  <i
    class="fr-rating-item icon-star-full"
    role="button"
    tabindex="0"
  />
  <i
    class="fr-rating-item icon-star-full"
    role="button"
    tabindex="0"
  />
  <i
    class="fr-rating-item icon-star-empty"
    role="button"
    tabindex="0"
  />
  <i
    class="fr-rating-item icon-star-empty"
    role="button"
    tabindex="0"
  />
</div>
`;

exports[`Testing functionalities 2`] = `
<DocumentFragment>
  <div
    class="fr-rating-container "
  >
    <i
      class="fr-rating-item icon-star-full"
      role="button"
      tabindex="0"
    />
    <i
      class="fr-rating-item icon-star-full"
      role="button"
      tabindex="0"
    />
    <i
      class="fr-rating-item icon-star-full"
      role="button"
      tabindex="0"
    />
    <i
      class="fr-rating-item icon-star-empty"
      role="button"
      tabindex="0"
    />
    <i
      class="fr-rating-item icon-star-empty"
      role="button"
      tabindex="0"
    />
  </div>
</DocumentFragment>
`;
